File: CSES/2. Sorting and Searching/CSES-1621_Distinct Numbers.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

void solve1() {
    int n; cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    sort(nums.begin(), nums.end());
    int ans = 1;
    for (int i = 1; i < n; i++) {
        if (nums[i] != nums[i - 1]) ans++;
    }
    cout << ans << endl;
}

void solve2() {
    int n, x; 
    cin >> n;
    set<int> nums;
    for (int i = 0; i < n; i++) {
        cin >> x;
        nums.insert(x);
    }
    cout << nums.size() << endl;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    // solve1();
    solve2();
    return 0;
}

File: CSES/2. Sorting and Searching/CSES-1621_Distinct Numbers.py
""" CSES-1621 Distinct Numbers
    Sort / Hash Set
    這題的測資中有針對Hash Collision的，會讓 set 的時間複雜度從 O(1) 變成 O(n) ，因此會 TLE 。
    可以透過自訂 hash function 來解決，這邊是直接改用字串來作為 hash key
"""
n = int(input())
nums = list(map(int, input().split()))

def solve1():
    nums.sort()
    ans = 1
    for i in range(1, n):
        if nums[i] != nums[i-1]:
            ans += 1
    print(ans)

def solve2():
    hash_set = set()
    for num in nums:
        hash_set.add(str(num)) # 改用字串來作為 hash key
    print(len(hash_set))

# solve1()
solve2()